---
title: "HW6"
author: "Kate Harline"
date: "4/5/2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Problem 1
# A

The assumption of a linear regression is that $\epsilon \sim N(0, \sigma^2)$ regardless of the distribution of the independent variables (genotypes, X). This imposes the same assumption on the dependent variable (phenotypes, Y) because in the simplest case for one individual $\hat{y_i} = \hat{\beta_{\mu}} + x_{i,a}\hat{\beta_{a}} + x_{i,d}\hat{\beta_{d}} + \epsilon$ when a polymorphism is noncausal,  $\hat{\beta_{a}} = \hat{\beta_{d}} = 0$ so $\hat{y_i} = \hat{\beta_{\mu} + \epsilon \sim N(0, \sigma^2)$ therefore the distribution for the linear regression model follows $Pr(Y | X) \sim N(\beta_{\mu} + X_{a}{\beta_{a}} + X_{d}\beta_{d}, \sigma^2_{\epsilon}$.

# B

Correctly rejecting the null hypothesis in a GWAS signifies that the region that contains this "hit" contains the causal polymorphism, not that the "hit" itself is causal. Due to the structure of genomes and the current depth of sequencing data, a GWAS is a mapping experiment. Genomic structure and inheritance patterns dictate that through linkage disequilibrium, sites that are closer together are inherited together. 

On the other hand, there is always a chance the null hypothesis was rejected incorrectly and therefore we have obtained a false positive. We control for the false positives by setting a lower alpha and doing multi-test corrections. 

## Problem 2

# A
```{r}
# input the phenotype data and report the number of samples, s
phenotypes <- read.table("./QG19 - hw6_phenotypes.txt", header = T)
n <- length(phenotypes[,1])
cat('There are n = ', n, 'samples')
```

# B
```{r}
# make a histogram of the phenotype data
hist(phenotypes[,1], main = 'Distribution of Scaled Heights', xlab = 'Scaled Height')
```

# C
```{r}
# input genotype data, report the number of genotypes N, sample size n
genotypes <- read.table("./QG19 - hw6_genotypes.txt", header = T)
N <- ncol(genotypes)

n <- nrow(genotypes) / 2
cat('There are N = ', N, ' genotypes. There are n = ', n, ' samples.')
```

# D
```{r}
# convert genotypes into xa and xd matrices
# set up new matrix
xa_matrix <- matrix(NA, nrow = nrow(genotypes), ncol = ncol(genotypes))
# fill with x_a values
for (i in 1:ncol(genotypes)) {
  xa_matrix[,i] <- sapply(genotypes[,i], (function(x) switch (x,
    'A1A1' = -1, 'A1A2' = 0, 'A2A2' = 1
  )))
}
# then convert x_a to x_d
xd_matrix <- -2*abs(xa_matrix)+1

```

# E
```{r}
# calculate MLE of b's, calculate f statistic, 
# install.packages(MASS)
library(MASS)
# Function to calculate the pval given a set of individuals' phenotype, and genotype encodings.
pval_calculator <- function(pheno_input, xa_input, xd_input){
    n_samples <- length(xa_input)

    X_mx <- cbind(1,xa_input,xd_input)
    
    MLE_beta <- ginv(t(X_mx) %*% X_mx) %*% t(X_mx) %*% pheno_input
    y_hat <- X_mx %*% MLE_beta
  
    SSM <- sum((y_hat - mean(pheno_input))^2)
    SSE <- sum((pheno_input - y_hat)^2)
  
    df_M <- 2
    df_E <- n_samples - 3 
  
    MSM <- SSM / df_M
    MSE <- SSE / df_E
    
    Fstatistic <- MSM / MSE
  
    pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)

    return(pval)
}

# Initialize some variables and constants
n_geno <- ncol(xa_matrix)
n_pheno <- length(phenotypes)
pvals <- c()

# Calculate and save pvals for each phenotype-genotype pair

  for (i in 1 : n_geno){
      pvals <- c(pvals, pval_calculator(phenotypes, 
                                      xa_input = xa_matrix[,i], 
                                      xd_input = xd_matrix[,i]))
  }


```

# F
```{r}
# create Manhattan plot of the data
library(ggplot2)
library(gridExtra)

neg_log_p <- -1 * log(pvals)
x <- seq(1, length(neg_log_p))
y <- neg_log_p

plot1 <-ggplot(data = data.frame(x, y), mapping = aes(x, y)) + geom_point() + geom_hline(yintercept=-log(0.05), col='blue') + labs(x = 'Genomic Position', y= '-log(p)', title = 'GWA Hits') + ylim(0,30)

print(plot1)
```

# G
```{r}
# create QQ plot of the data
# plot expected v observed
obs <- sort(-log10(pvals))
# generate range
exp <- seq(from = (1 / length(obs)), to = 1, length.out = length(obs))
# - log 10
exp <- sort(-log10(exp))

plot1 <-ggplot(data = data.frame(exp, obs), mapping = aes(exp, obs)) + geom_point() +  labs(title = 'QQ Plot for Scaled Heights') 
```

# H
The QQ plot for the data shows ...

# I 
```{r}
# use bonferroni correction to reject the null with alpha = 0.5, 
r <- sum(pvals < 0.05/ncol(xa_matrix))

cat('The null was rejected ', r, ' times.')
```

# J 
By chance, given the number of genotypes measured, we would expect $1194 * 0.05 \approx 60 $ hits. So ....

## Problem 3